src/quarot/model_adapter.py

```diff
@@ -0,0 +1,461 @@
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_attention_inputs(self) -> Sequence[Linear]:
+        """
+        Returns a list of the Linear layers (nn.modules) that are inputs to the attention component.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_attention_output(self) -> Linear:
+        """
+        Returns the Linear layer (nn.module) that is the output of the attention component.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_mlp_inputs(self) -> Sequence[Linear]:
+        """
+        Returns a list of the Linear layers (nn.modules) that are inputs to the MLP component.
+        For simple mlps, this will be a list of length 1. For gated mlps (as in the Llama models) there will be two.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_mlp_output(self) -> Linear:
+        """
+        Returns the Linear layer (nn.module) that is the output of the MLP component (usually fc2 or down_proj)
+        """
+        raise NotImplementedError
+
+    def get_updated_args(self, hidden_states: Any, args: tuple) -> tuple:
+        """
+        `args` is a tuple of the arguments to the layer's forward method. hidden_states is the new value for the
+        hidden_states argument. This method returns a new tuple of arguments with the hidden_states argument updated.
+        """
+        return (
+            args[: self.hidden_states_args_position] + (hidden_states,) + args[self.hidden_states_args_position + 1 :]
+        )
+
+    # The below are QuaRot specific. Consider renaming these to something more general if non Llama-2 architectures name
+    # these differently.
+    @abstractmethod
+    def get_v_proj(self) -> Linear:
+        """
+        Returns the Linear layer (nn.module) that is the v projection in the attention component.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_o_proj(self) -> Linear:
+        """
+        Returns the Linear layer (nn.module) that is the output projection in the attention component.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_rope_function_name(self) -> str:
+        """
+        Returns the name of the rotary positional embedding (RoPE) function in the layer.
+        TODO: This is Llama-2 specific, how to generalize?
+        """
+        raise NotImplementedError
+
+
+class ModelAdapter(ABC):
+    """
+    To implement a new model adapter, implement the interface defined in this class.
+    """
+
+    def __init__(self):
+        self.quarot_conf: QuaRotConfig | None = None
+
+    @property
+    @abstractmethod
+    def model(self) -> Module:
+        """
+        The original model that quarot interacts with.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def config(self) -> object:
```

Comment:
The reason we introduced model adapters was that there would be a clear and explicit interface for model transformation code to use and for people wanting to transform their models to implement. Having an untyped property in the model adapter interface defeats that purpose, because the model transformation code can not meaningfully and reliably use such properties, while implementers of the interface have no way of knowing what structure they should return here for the transformation code to work.

Reply:
Interesting. I was not there, when this property made it into slicegpt. It seems unused there.
Here, in quarot, it competes with the `hidden_size` property. I suggest keeping the `hidden_size` and adding a `num_attention_heads` properties instead of having the `config` property, because these two are what we actually need for our algorithms.

Reply:
If we are just changing the return type from `object` to `PretrainedConfig` for the `config` property, I think that's fine.